packages/rest/src/validation/request-body.validator.ts

```diff
@@ -4,21 +4,28 @@
 // License text available at https://opensource.org/licenses/MIT
 
 import {
+  ReferenceObject,
   RequestBodyObject,
   SchemaObject,
-  ReferenceObject,
   SchemasObject,
 } from '@loopback/openapi-v3-types';
 import * as AJV from 'ajv';
 import * as debugModule from 'debug';
-import * as util from 'util';
-import {HttpErrors, RestHttpErrors, RequestBody} from '..';
 import * as _ from 'lodash';
+import * as util from 'util';
+import {HttpErrors, RequestBody, RestHttpErrors} from '..';
 
 const toJsonSchema = require('openapi-schema-to-json-schema');
 const debug = debugModule('loopback:rest:validation');
 
-export type RequestBodyValidationOptions = AJV.Options;
+export interface RequestBodyValidationOptions extends AJV.Options {
+  /**
+   * Skip cache for AJV validators. This flag is useful to test different
+   * AJV options as the caching key is the SchemaObject/ReferenceObject without
+   * honoring options.
+   */
+  skipCache?: boolean;
```

Comment:
What happens if the application configures two body parsers with different validation options? For example, `JSON` parser allowing nullable fields, `urlencoded-form` parser not allowing nullable fields. IIUC, the first request will populate the cache with a precompiled validator, and then all subsequent requests will use that validator regardless of parser settings.

IMO, we need to fix the caching mechanism to include (custom) validation options in the key.

Reply:
We have only one `RequestBodyValidationOptions` from `RequestBodyParserOptions`. The flag is introduced mostly for testing. In production environment, the `options` is fixed based on current code.

Reply:
This is what the documentation says - emphasis is mine:

The request body parser options is bound to `RestBindings.REQUEST_BODY_PARSER_OPTIONS`. To customize request body parser
options, you can simply bind a new value to its key.

Built-in parsers retrieve their own options from the request body parser options. **The parser specific properties override common ones.** For example, given
the following configuration:
```ts
{
  limit: '1MB'
  json: {
    strict: false
  },
  text: {
    limit: '2MB'
  },
  // Validation options for AJV, see https://github.com/epoberezkin/ajv#options
  validation: {nullable: true},
}
```

---

As I understand that, it is possible to customize validation options on per-parser basis, and thus my original comment still applies.

PTAL.

To be honest, I find configurable AJV rules as an overkill that's not needed at this point.

> The flag is introduced mostly for testing. 

I understand that in tests, you don't want the cache to be preserved across individual tests. Let's find a better way how to accomplish that. 

Few examples:
1. We can include important options fields when computing the cache key, so that we have different cached validators for different combination of options.
2. We can expose a new internal function to clean `compiledSchemaCache`, this function will be called in tests from a `beforeEach` hook
3. We can change `compiledSchemaCache` from an internal singleton to a new application-level binding. Then in tests, whenever we create a new application instance, it will start with an empty cache. Tests sharing the same app instance can re-bind the cache.

Reply:
I changed the code to allow a weakmap to be passed in via `options` and it falls back to the built-in singleton.

Reply:
> I changed the code to allow a weakmap to be passed in via options and it falls back to the built-in singleton.

I am fine with using this approach to enable the tests 👍

Reply:
Having said that, I am not convinced that it addresses my concern about parser-specific options being ignored because of caching. Please a tests to verify, e.g. using the example I posted above for inspiration:

> For example, `JSON` parser allowing nullable fields, `urlencoded-form` parser not allowing nullable fields. IIUC, the first request will populate the cache with a precompiled validator, and then all subsequent requests will use that validator regardless of parser settings.